### PR TITLE
Refactor mixed_semi_join using cuco::static_set

### PR DESCRIPTION
**Pull request by srinivasyadav18**
## Description
This PR refactors `mixed_semi_join` by replacing **cuco** legacy `static_map` with latest `static_set`. 
Contributes to #12261.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/join/mixed_join_semi.cu [hunk 1]
```diff
@@ -197,25 +152,33 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
   auto const equality_build_conditional =
     row_comparator_conditional_build.equal_to<false>(build_nulls, compare_nulls);
   double_row_equality equality_build{equality_build_equality, equality_build_conditional};
-  make_pair_function_semi pair_func_build{};
 
-  auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func_build);
+  hash_set_type set{compute_hash_table_size(build.num_rows()),
```
Comment (PointKernel) [inline cpp/src/join/mixed_join_semi.cu:156]:
```suggestion
  hash_set_type set{build.num_rows(),
```
Using `compute_hash_table_size` together with `0.5` will make the hash table with 0.25 occupancy.

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 2]
```diff
@@ -69,8 +69,8 @@ __attribute__((visibility("hidden"))) __launch_bounds__(block_size) __global__
     auto equality = single_expression_equality<has_nulls>{
       evaluator, thread_intermediate_storage, false, equality_probe};
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+    auto set_ref_ = set_ref.with_hash_function(hash_probe).with_key_eq(equality);
```
Comment (PointKernel) [inline cpp/src/join/mixed_join_kernels_semi.cu:72]: we probably want to pass the custom key/hash on the host to save some time

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 3]
```diff
@@ -160,6 +161,42 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
...
+  row_equality _equality_comparator;
+  row_equality _conditional_comparator;
+};
+
+// The hash table type used by mixed_semi_join with the build_table.
+using hash_set_type = cuco::static_set<hash_value_type,
+                                       cuco::extent<size_t>,
+                                       cuda::thread_scope_device,
+                                       double_row_equality,
+                                       cuco::linear_probing<1, row_hash>,
```
Comment (PointKernel) [inline cpp/src/join/mixed_join_common_utils.cuh:193]: Have you tuned CG size like 2 or 4 as well as the probing scheme like double hashing?

Comment (srinivasyadav18) [reply cpp/src/join/mixed_join_common_utils.cuh]: I tried using CG 2, 4. And CG with 4 gave closer performance to current implementation with static_map. 
**double hashing** is not supported with **static_set** `.with_hash_function` API yet.

File: cpp/src/join/mixed_join_kernels_semi.cu (hunk 2 above)
Comment (srinivasyadav18) [reply cpp/src/join/mixed_join_kernels_semi.cu]: Resolved in [dc7b7d1](https://github.com/rapidsai/cudf/pull/16230/commits/dc7b7d1620c5fbebc24fe61a4aef89b03176d048)

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 4]
```diff
@@ -70,8 +71,9 @@ CUDF_HIDDEN __launch_bounds__(block_size) __global__
     auto equality = single_expression_equality<has_nulls>{
       evaluator, thread_intermediate_storage, false, equality_probe};
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+    auto set_ref_equality = set_ref.with_key_eq(equality);
+    const auto result     = set_ref_equality.contains(tile, outer_row_index);
```
Comment (JayjeetAtGithub) [inline cpp/src/join/mixed_join_kernels_semi.cu:75]: Maybe we can use east const here

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 5]
```diff
@@ -54,13 +58,10 @@ CUDF_HIDDEN __launch_bounds__(block_size) __global__
...
-    &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
-
-  cudf::size_type const left_num_rows  = left_table.num_rows();
-  cudf::size_type const right_num_rows = right_table.num_rows();
-  auto const outer_num_rows            = left_num_rows;
+    &intermediate_storage[(threadIdx.x / cg_size) * device_expression_data.num_intermediates];
 
-  cudf::size_type outer_row_index = threadIdx.x + blockIdx.x * block_size;
+  cudf::size_type const outer_num_rows = left_table.num_rows();
+  cudf::size_type outer_row_index      = (threadIdx.x + blockIdx.x * block_size) / cg_size;
```
Comment (JayjeetAtGithub) [inline cpp/src/join/mixed_join_kernels_semi.cu:64]: Can we also make `outer_row_index` a const ?

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 6]
```diff
@@ -70,8 +71,9 @@ CUDF_HIDDEN __launch_bounds__(block_size) __global__
     auto equality = single_expression_equality<has_nulls>{
       evaluator, thread_intermediate_storage, false, equality_probe};
 
-    left_table_keep_mask[outer_row_index] =
-      hash_table_view.contains(outer_row_index, hash_probe, equality);
+    auto set_ref_equality = set_ref.with_key_eq(equality);
```
Comment (JayjeetAtGithub) [inline cpp/src/join/mixed_join_kernels_semi.cu:74]: Can this also be a const?

**Review by JayjeetAtGithub — CHANGES_REQUESTED** [review]

**Comment by ttnghia** [comment]
> This PR refactors `mixed_semi_join` by replacing **cuco** legacy `static_map` with latest `static_set`. 

Please provide benchmark showing what benefit this PR will bring to us.

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 7]
```diff
@@ -160,6 +161,42 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
   }
 };
 
+/**
+ * @brief Equality comparator that composes two row_equality comparators.
+ */
+class double_row_equality {
+ public:
+  double_row_equality(row_equality equality_comparator, row_equality conditional_comparator)
+    : _equality_comparator{equality_comparator}, _conditional_comparator{conditional_comparator}
```
Comment (ttnghia) [inline cpp/src/join/mixed_join_common_utils.cuh:170]: Use `std::move` on the parameters to avoid copying.

Comment (ttnghia) [reply cpp/src/join/mixed_join_common_utils.cuh:170]: Indeed we can use `struct` with all public members so we won't even need a constructor.

File: cpp/src/join/mixed_join_semi.cu [hunk 8]
```diff
@@ -197,31 +152,43 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
   auto const equality_build_conditional =
     row_comparator_conditional_build.equal_to<false>(build_nulls, compare_nulls);
   double_row_equality equality_build{equality_build_equality, equality_build_conditional};
-  make_pair_function_semi pair_func_build{};
 
-  auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func_build);
+  auto const build_num_rows = compute_hash_table_size(build.num_rows());
+
+  hash_set_type set{build_num_rows,
```
Comment (ttnghia) [inline cpp/src/join/mixed_join_semi.cu:158]: In general, avoid using standard keywords as name like this. So instead of `set`, we can call it `row_set` or else.
```suggestion
  hash_set_type row_set{build_num_rows,
```

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 9]
```diff
@@ -160,6 +161,42 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
...
+           _conditional_comparator(lhs_index_type{lhs_row_index}, rhs_index_type{rhs_row_index});
+  }
+
+ private:
+  row_equality _equality_comparator;
+  row_equality _conditional_comparator;
+};
+
+// The hash table type used by mixed_semi_join with the build_table.
+using hash_set_type = cuco::static_set<size_type,
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_common_utils.cuh:182]: Used `hash_set_type` instead of the suggested for consistency across cudf.

File: cpp/src/join/mixed_join_common_utils.cuh [hunk 10]
```diff
@@ -160,6 +161,42 @@ struct pair_expression_equality : public expression_equality<has_nulls> {
...
+                                       cuco::extent<size_t>,
+                                       cuda::thread_scope_device,
+                                       double_row_equality,
+                                       cuco::linear_probing<4, row_hash>,
+                                       cudf::detail::cuco_allocator,
+                                       cuco::storage<1>>;
+
+// The device_ref_type used by mixed_semi_join kerenels for probing.
+using hash_set_ref_type = hash_set_type::ref_type<cuco::contains_tag>;
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_common_utils.cuh:191]: Same reason as above

File: cpp/src/join/mixed_join_semi.cu [hunk 11]
```diff
@@ -160,22 +121,16 @@ std::unique_ptr<rmm::device_uvector<size_type>> mixed_join_semi(
   auto const preprocessed_probe =
     experimental::row::equality::preprocessed_table::create(probe, stream);
   auto const row_comparator =
-    cudf::experimental::row::equality::two_table_comparator{preprocessed_probe, preprocessed_build};
+    cudf::experimental::row::equality::two_table_comparator{preprocessed_build, preprocessed_probe};
```
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_semi.cu]: I need to dig into why but without this change the gtests start failing. Maybe @srinivasyadav18 can provide us a quick insight.

**Comment by mhaseeb123** [comment]
> > This PR refactors `mixed_semi_join` by replacing **cuco** legacy `static_map` with latest `static_set`.
> 
> Please provide benchmark showing what benefit this PR will bring to us.

```
# Benchmark Results

## mixed_left_semi_join

### [0] NVIDIA RTX 5880 Ada Generation

| Key | Nullable | left_size | right_size |  GPU Time_old  |  GPU Time_new  |  pct_change  |
|-----|----------|-----------|------------|----------------|----------------|--------------|
| I32 |        0 |      1000 |       1000 |      97.414 us |      91.512 us |       6.059% |
| I32 |        0 |    100000 |       1000 |     136.931 us |     141.995 us |      -3.698% |
| I32 |        0 |  10000000 |       1000 |       1.495 ms |       1.378 ms |       7.826% |
| I32 |        0 |    100000 |     100000 |     160.755 us |     157.231 us |       2.192% |
| I32 |        0 |  10000000 |     100000 |       1.486 ms |       1.434 ms |       3.499% |
| I32 |        0 |  10000000 |   10000000 |       5.868 ms |       3.291 ms |      43.916% |
| I32 |        1 |      1000 |       1000 |     111.645 us |     110.318 us |       1.189% |
| I32 |        1 |    100000 |       1000 |     147.585 us |     157.790 us |      -6.915% |
| I32 |        1 |  10000000 |       1000 |       1.322 ms |       1.414 ms |      -6.959% |
| I32 |        1 |    100000 |     100000 |     163.264 us |     160.227 us |        1.86% |
| I32 |        1 |  10000000 |     100000 |       1.133 ms |       1.377 ms |     -21.536% |
| I32 |        1 |  10000000 |   10000000 |       2.970 ms |       2.530 ms |      14.815% |
| I64 |        0 |      1000 |       1000 |      98.314 us |      93.307 us |       5.093% |
| I64 |        0 |    100000 |       1000 |     138.418 us |     142.360 us |      -2.848% |
| I64 |        0 |  10000000 |       1000 |       1.474 ms |       1.397 ms |       5.224% |
| I64 |        0 |    100000 |     100000 |     162.779 us |     158.851 us |       2.413% |
| I64 |        0 |  10000000 |     100000 |       1.513 ms |       1.447 ms |       4.362% |
| I64 |        0 |  10000000 |   10000000 |       6.871 ms |       4.276 ms |      37.767% |
| I64 |        1 |      1000 |       1000 |     111.365 us |     109.864 us |       1.348% |
| I64 |        1 |    100000 |       1000 |     148.109 us |     157.547 us |      -6.372% |
| I64 |        1 |  10000000 |       1000 |       1.170 ms |       1.429 ms |     -22.137% |
| I64 |        1 |    100000 |     100000 |     162.729 us |     162.550 us |        0.11% |
| I64 |        1 |  10000000 |     100000 |       1.163 ms |       1.392 ms |      -19.69% |
| I64 |        1 |  10000000 |   10000000 |       3.033 ms |       2.674 ms |      11.836% |
```

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 12]
```diff
@@ -52,24 +56,24 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
   auto evaluator = cudf::ast::detail::expression_evaluator<has_nulls>(
     left_table, right_table, device_expression_data);
 
   if (outer_row_index < outer_num_rows) {
+    // Make sure to swap_tables here as hash_set will use probe table as the left one.
+    auto constexpr swap_tables = true;
     // Figure out the number of elements for this key.
     auto equality = single_expression_equality<has_nulls>{
-      evaluator, thread_intermediate_storage, false, equality_probe};
+      evaluator, thread_intermediate_storage, swap_tables, equality_probe};
```
Comment (mhaseeb123) [inline cpp/src/join/mixed_join_kernels_semi.cu:72]: @karthikeyann

File: cpp/src/join/mixed_join_semi.cu (hunk 11 above)
[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_semi.cu:124]: Update: Shouldn't have changed the order here and instead needed to set `swap_tables=true` at mixed_join_kernels_semi.cu:+72. This is because we reset the hasher and equality functors for the hash_set swapping the `left` and `right` directions of the comparisons.

[the remark this answers is not in the document]
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_semi.cu:124]: Doing the swap here was just putting a patch on the underlying problem which was being caught in Java tests. I replicated the problem in C++ tests and have added the new test too so it can be caught early on.

File: docs/cudf/source/cudf_pandas/how-it-works.md [hunk 13]
```diff
@@ -44,20 +44,11 @@ allocation may be a bottleneck depending on the workload. Managed memory
...
 Other memory allocators can be used by changing the environment
-variable `CUDF_PANDAS_RMM_MODE` to one of the following:
+variable `CUDF_PANDAS_RMM_MODE` to one of the following.
 
-1. `"managed_pool"` (default, if supported): CUDA Unified Memory (managed memory) with RMM's asynchronous pool allocator.
-2. `"managed"`: CUDA Unified Memory, (managed memory) with no pool allocator.
-3. `"async"`: CUDA's built-in pool asynchronous pool allocator with normal CUDA device memory.
-4. `"pool"` (default if `"managed_pool"` is not supported): RMM's asynchronous pool allocator with normal CUDA device memory.
-5. `"cuda"`: normal CUDA device memory with no pool allocator.
+1. "managed_pool" (default): CUDA Unified Memory (managed memory) with RMM's asynchronous pool allocator.
```
Comment (mhaseeb123) [inline docs/cudf/source/cudf_pandas/how-it-works.md:50]: Should we revert the removed ``s here?

File: cpp/src/join/mixed_join_common_utils.cuh (hunk 3 above)
Comment (mhaseeb123) [reply cpp/src/join/mixed_join_common_utils.cuh]: @srinivasyadav18 by closer performance, do you mean closer but slower or closer or faster? What about cg_size=1? I can probably benchmark and tune cg_size again but would help if you have any data.

Comment (mhaseeb123) [reply cpp/src/join/mixed_join_common_utils.cuh]: Update: I tried with cg_size=1,2,4 and 4 seems to be the best among them but only marginally so going with that.

File: python/cudf/cudf/pandas/__init__.py [hunk 14]
```diff
@@ -25,8 +25,8 @@
 }
 
 
-def _enable_managed_prefetching(rmm_mode, managed_memory_is_supported):
-    if managed_memory_is_supported and "managed" in rmm_mode:
+def _enable_managed_prefetching(rmm_mode):
```
Comment (mhaseeb123) [inline python/cudf/cudf/pandas/__init__.py:28]: @srinivasyadav18 Are these changes intentional?

File: docs/cudf/source/cudf_pandas/how-it-works.md (hunk 13 above)
Comment (mhaseeb123) [reply docs/cudf/source/cudf_pandas/how-it-works.md:50]: @srinivasyadav18 Are these changes intentional?

File: python/cudf/cudf/pandas/__init__.py (hunk 14 above)
Comment (mhaseeb123) [reply python/cudf/cudf/pandas/__init__.py]: Assuming these are stale files and reverting them.

File: docs/cudf/source/cudf_pandas/how-it-works.md (hunk 13 above)
Comment (srinivasyadav18) [reply docs/cudf/source/cudf_pandas/how-it-works.md]: @mhaseeb123 No. Sorry, these came up from rebasing.

File: cpp/src/join/mixed_join_kernels_semi.cu (hunk 12 above)
Comment (karthikeyann) [reply cpp/src/join/mixed_join_kernels_semi.cu:72]: 👍

File: cpp/src/join/mixed_join_kernels_semi.cu [hunk 15]
```diff
@@ -52,24 +56,24 @@ CUDF_KERNEL void __launch_bounds__(block_size)
...
-    &intermediate_storage[threadIdx.x * device_expression_data.num_intermediates];
-
-  cudf::size_type const left_num_rows  = left_table.num_rows();
-  cudf::size_type const right_num_rows = right_table.num_rows();
-  auto const outer_num_rows            = left_num_rows;
+    &intermediate_storage[tile.meta_group_rank() * device_expression_data.num_intermediates];
 
-  cudf::size_type outer_row_index = threadIdx.x + blockIdx.x * block_size;
+  cudf::size_type const outer_num_rows  = left_table.num_rows();
+  cudf::size_type const outer_row_index = (threadIdx.x + blockIdx.x * block_size) / cg_size;
```
Comment (PointKernel) [inline cpp/src/join/mixed_join_kernels_semi.cu:62]:
```suggestion
  cudf::size_type const outer_num_rows  = left_table.num_rows();
  auto const outer_row_index = cudf::detail::grid_1d::global_thread_id<block_size>() / cg_size;
```
https://github.com/rapidsai/cudf/blob/3dbc33a5cb1cf7052cd67f5654b34594403fbfef/cpp/include/cudf/detail/utilities/cuda.cuh#L91-L143
Ideally, we need to always use `grid_1d` to calculate all global thread indices and strides.

File: cpp/src/join/mixed_join_semi.cu (hunk 11 above)
[the remark this answers is not in the document]
Comment (PointKernel) [reply cpp/src/join/mixed_join_semi.cu:124]: This is mainly due to a breaking change in https://github.com/NVIDIA/cuCollections/pull/479 where we always put query keys on the left-hand side for comparison.

**Review by PointKernel — APPROVED** [review]
One minor improvement otherwise looks great to me.

@mhaseeb123 Thank you for stepping in to help resolve the issue and finalize the PR! :fire: :fire: :fire:

**Comment by mhaseeb123** [comment]
Performance results from Spark (spark-h). Thanks to @zpuller for running this. Queries of interest: q16 and q94 that use `mixed_semi_join`

```
Regression alerts
-----------------

Speedup results
-----------------
query1: Previous (1881.6 ms) vs Current (1774.4 ms) Diff 107 E2E 1.06x
query2: Previous (1651.8 ms) vs Current (1693.4 ms) Diff -41 E2E 0.98x
query3: Previous (441.0 ms) vs Current (440.2 ms) Diff 0 E2E 1.00x
query4: Previous (8464.8 ms) vs Current (8578.4 ms) Diff -113 E2E 0.99x
query5: Previous (1917.6 ms) vs Current (2090.8 ms) Diff -173 E2E 0.92x
query6: Previous (1026.4 ms) vs Current (946.8 ms) Diff 79 E2E 1.08x
query7: Previous (2728.8 ms) vs Current (2871.4 ms) Diff -142 E2E 0.95x
query8: Previous (1070.8 ms) vs Current (1062.4 ms) Diff 8 E2E 1.01x
query9: Previous (3184.8 ms) vs Current (3115.6 ms) Diff 69 E2E 1.02x
query10: Previous (1529.4 ms) vs Current (1480.2 ms) Diff 49 E2E 1.03x
query11: Previous (4687.8 ms) vs Current (4840.4 ms) Diff -152 E2E 0.97x
query12: Previous (552.2 ms) vs Current (565.8 ms) Diff -13 E2E 0.98x
query13: Previous (1100.6 ms) vs Current (1098.0 ms) Diff 2 E2E 1.00x
query14_part1: Previous (5230.6 ms) vs Current (5320.0 ms) Diff -89 E2E 0.98x
query14_part2: Previous (4355.2 ms) vs Current (4310.8 ms) Diff 44 E2E 1.01x
query15: Previous (1063.8 ms) vs Current (1098.4 ms) Diff -34 E2E 0.97x
query16: Previous (3262.6 ms) vs Current (3335.4 ms) Diff -72 E2E 0.98x
query17: Previous (1844.0 ms) vs Current (1744.4 ms) Diff 99 E2E 1.06x
query18: Previous (3182.2 ms) vs Current (3178.6 ms) Diff 3 E2E 1.00x
query19: Previous (1488.0 ms) vs Current (1500.8 ms) Diff -12 E2E 0.99x
query20: Previous (591.6 ms) vs Current (629.8 ms) Diff -38 E2E 0.94x
query21: Previous (548.4 ms) vs Current (635.4 ms) Diff -87 E2E 0.86x
query22: Previous (962.0 ms) vs Current (950.6 ms) Diff 11 E2E 1.01x
query23_part1: Previous (10493.8 ms) vs Current (10460.0 ms) Diff 33 E2E 1.00x
query23_part2: Previous (14016.4 ms) vs Current (14153.8 ms) Diff -137 E2E 0.99x
query24_part1: Previous (6507.8 ms) vs Current (6394.2 ms) Diff 113 E2E 1.02x
query24_part2: Previous (6494.6 ms) vs Current (6488.6 ms) Diff 6 E2E 1.00x
query25: Previous (1638.2 ms) vs Current (1599.6 ms) Diff 38 E2E 1.02x
query26: Previous (849.2 ms) vs Current (851.0 ms) Diff -1 E2E 1.00x
query27: Previous (1346.0 ms) vs Current (1222.6 ms) Diff 123 E2E 1.10x
query28: Previous (6245.0 ms) vs Current (5135.8 ms) Diff 1109 E2E 1.22x
query29: Previous (2420.2 ms) vs Current (2526.8 ms) Diff -106 E2E 0.96x
query30: Previous (1756.0 ms) vs Current (1744.6 ms) Diff 11 E2E 1.01x
query31: Previous (2492.2 ms) vs Current (2522.8 ms) Diff -30 E2E 0.99x
query32: Previous (833.0 ms) vs Current (874.2 ms) Diff -41 E2E 0.95x
query33: Previous (1374.6 ms) vs Current (1313.0 ms) Diff 61 E2E 1.05x
query34: Previous (1649.6 ms) vs Current (1662.8 ms) Diff -13 E2E 0.99x
query35: Previous (1948.6 ms) vs Current (1984.4 ms) Diff -35 E2E 0.98x
query36: Previous (965.0 ms) vs Current (1033.8 ms) Diff -68 E2E 0.93x
query37: Previous (1196.2 ms) vs Current (1344.4 ms) Diff -148 E2E 0.89x
query38: Previous (3036.8 ms) vs Current (3376.4 ms) Diff -339 E2E 0.90x
query39_part1: Previous (2122.8 ms) vs Current (2113.8 ms) Diff 9 E2E 1.00x
query39_part2: Previous (1412.2 ms) vs Current (1311.0 ms) Diff 101 E2E 1.08x
query40: Previous (1165.6 ms) vs Current (1141.0 ms) Diff 24 E2E 1.02x
query41: Previous (326.4 ms) vs Current (306.0 ms) Diff 20 E2E 1.07x
query42: Previous (318.2 ms) vs Current (312.2 ms) Diff 6 E2E 1.02x
query43: Previous (726.8 ms) vs Current (739.4 ms) Diff -12 E2E 0.98x
query44: Previous (624.8 ms) vs Current (637.8 ms) Diff -13 E2E 0.98x
query45: Previous (1137.0 ms) vs Current (1100.4 ms) Diff 36 E2E 1.03x
query46: Previous (1616.2 ms) vs Current (1655.2 ms) Diff -39 E2E 0.98x
query47: Previous (1465.8 ms) vs Current (1507.8 ms) Diff -42 E2E 0.97x
query48: Previous (839.0 ms) vs Current (825.4 ms) Diff 13 E2E 1.02x
query49: Previous (2236.4 ms) vs Current (2186.2 ms) Diff 50 E2E 1.02x
query50: Previous (6857.4 ms) vs Current (7052.4 ms) Diff -195 E2E 0.97x
query51: Previous (2435.4 ms) vs Current (2233.2 ms) Diff 202 E2E 1.09x
query52: Previous (393.0 ms) vs Current (421.4 ms) Diff -28 E2E 0.93x
query53: Previous (732.2 ms) vs Current (562.8 ms) Diff 169 E2E 1.30x
query54: Previous (1319.2 ms) vs Current (1375.2 ms) Diff -56 E2E 0.96x
query55: Previous (363.8 ms) vs Current (368.8 ms) Diff -5 E2E 0.99x
query56: Previous (969.4 ms) vs Current (927.2 ms) Diff 42 E2E 1.05x
query57: Previous (1292.4 ms) vs Current (1290.6 ms) Diff 1 E2E 1.00x
query58: Previous (1091.0 ms) vs Current (1014.8 ms) Diff 76 E2E 1.08x
query59: Previous (1601.6 ms) vs Current (1399.4 ms) Diff 202 E2E 1.14x
query60: Previous (1569.8 ms) vs Current (1518.2 ms) Diff 51 E2E 1.03x
query61: Previous (1357.6 ms) vs Current (1431.0 ms) Diff -73 E2E 0.95x
query62: Previous (901.8 ms) vs Current (921.0 ms) Diff -19 E2E 0.98x
query63: Previous (903.8 ms) vs Current (777.2 ms) Diff 126 E2E 1.16x
query64: Previous (13170.4 ms) vs Current (13072.8 ms) Diff 97 E2E 1.01x
query65: Previous (3076.6 ms) vs Current (3032.6 ms) Diff 44 E2E 1.01x
query66: Previous (2326.6 ms) vs Current (2304.2 ms) Diff 22 E2E 1.01x
query67: Previous (18200.0 ms) vs Current (18346.8 ms) Diff -146 E2E 0.99x
query68: Previous (1091.6 ms) vs Current (1146.6 ms) Diff -55 E2E 0.95x
query69: Previous (1468.6 ms) vs Current (1471.8 ms) Diff -3 E2E 1.00x
query70: Previous (1663.4 ms) vs Current (1521.0 ms) Diff 142 E2E 1.09x
query71: Previous (2684.6 ms) vs Current (2529.8 ms) Diff 154 E2E 1.06x
query72: Previous (2697.6 ms) vs Current (2676.8 ms) Diff 20 E2E 1.01x
query73: Previous (809.2 ms) vs Current (879.2 ms) Diff -70 E2E 0.92x
query74: Previous (3871.2 ms) vs Current (3882.0 ms) Diff -10 E2E 1.00x
query75: Previous (6280.0 ms) vs Current (6336.0 ms) Diff -56 E2E 0.99x
query76: Previous (2696.4 ms) vs Current (1886.0 ms) Diff 810 E2E 1.43x
query77: Previous (780.2 ms) vs Current (773.0 ms) Diff 7 E2E 1.01x
query78: Previous (7196.6 ms) vs Current (7067.0 ms) Diff 129 E2E 1.02x
query79: Previous (1078.2 ms) vs Current (1164.4 ms) Diff -86 E2E 0.93x
query80: Previous (3464.8 ms) vs Current (3399.6 ms) Diff 65 E2E 1.02x
query81: Previous (2228.0 ms) vs Current (2266.0 ms) Diff -38 E2E 0.98x
query82: Previous (2196.4 ms) vs Current (1903.0 ms) Diff 293 E2E 1.15x
query83: Previous (651.6 ms) vs Current (654.0 ms) Diff -2 E2E 1.00x
query84: Previous (1128.2 ms) vs Current (1030.4 ms) Diff 97 E2E 1.09x
query85: Previous (1634.2 ms) vs Current (1914.6 ms) Diff -280 E2E 0.85x
query86: Previous (897.4 ms) vs Current (922.0 ms) Diff -24 E2E 0.97x
query87: Previous (3044.8 ms) vs Current (2964.0 ms) Diff 80 E2E 1.03x
query88: Previous (3685.4 ms) vs Current (3625.6 ms) Diff 59 E2E 1.02x
query89: Previous (795.6 ms) vs Current (806.6 ms) Diff -11 E2E 0.99x
query90: Previous (986.8 ms) vs Current (844.8 ms) Diff 142 E2E 1.17x
query91: Previous (1151.2 ms) vs Current (1164.8 ms) Diff -13 E2E 0.99x
query92: Previous (456.2 ms) vs Current (512.0 ms) Diff -55 E2E 0.89x
query93: Previous (9548.6 ms) vs Current (9570.0 ms) Diff -21 E2E 1.00x
query94: Previous (3812.8 ms) vs Current (3756.2 ms) Diff 56 E2E 1.02x
query95: Previous (5580.8 ms) vs Current (5419.2 ms) Diff 161 E2E 1.03x
query96: Previous (3956.8 ms) vs Current (4010.6 ms) Diff -53 E2E 0.99x
query97: Previous (1694.6 ms) vs Current (1641.8 ms) Diff 52 E2E 1.03x
query98: Previous (1284.0 ms) vs Current (1339.4 ms) Diff -55 E2E 0.96x
query99: Previous (1526.4 ms) vs Current (1535.2 ms) Diff -8 E2E 0.99x
benchmark: Previous (273600.0 ms) vs Current (271000.0 ms) Diff 2600 E2E 1.01x
```

**Review by ttnghia — APPROVED** [review]

**Comment by ttnghia** [comment]
Shorter for easy reading:
```
Speedup results
-----------------
query16: Previous (3262.6 ms) vs Current (3335.4 ms) Diff -72 E2E 0.98x
query94: Previous (3812.8 ms) vs Current (3756.2 ms) Diff 56 E2E 1.02x
```
